Add flag for printing reports in colour

meshy opened the pull request:
With this new `--color` (or `--colour`, or `-c`) flag, the error report can now be printed in colour.

The selected colour scheme (mostly) matches that of Mypy. (I didn't bother adding URL highlighting in notes. Please let me know if you want this.)

I used ANSI colour codes for this to avoid adding a dependency, but please let me know if you would prefer this to make use of a library such as `rich` or `colorist`.

![image](https://github.com/Memrise/mypy-json-report/assets/767671/0be00216-b41d-4232-883d-f7db2341a3f8)
![image](https://github.com/Memrise/mypy-json-report/assets/767671/8fb61082-99fb-43c6-b26a-00c46a9dbfe2)
![image](https://github.com/Memrise/mypy-json-report/assets/767671/d19de9c0-9466-42ef-bd57-ad496a80193a)
![image](https://github.com/Memrise/mypy-json-report/assets/767671/212b5b2b-e6b3-4644-a6c3-13f5f205d24c)


Happy New Year!

meshy commented:
I've taken this out of review for a moment, because I intend to do a refactor PR first, and land this afterwards. That'll make it easier to revamp https://github.com/Memrise/mypy-json-report/pull/18

Tenzer reviewed (commented):
Happy new year Charlie!

Thank you for this, I've raised a few suggestions.

meshy commented on mypy_json_report.py:
```diff
@@ -229,6 +242,94 @@ class DiffReport:
+            location, _, message = line.partition("error:")
+            error_msg, _, code = message.rpartition("[")
+
+            if code:
```
Great catch, thanks!

meshy commented on mypy_json_report.py:
```diff
@@ -229,6 +242,94 @@ class DiffReport:
+class DefaultChangeReportWriter:
+    """Writes an error summary without color."""
+
+    def __init__(self, _write: Callable[[str], Any] = sys.stdout.write) -> None:
```
Yeah I did consider that, but as you said, I thought the function was more flexible.

meshy commented on mypy_json_report.py:
```diff
@@ -229,6 +242,94 @@ class DiffReport:
+        new_errors = "\n".join([self._format_line(line) for line in diff.error_lines])
+        self.write(new_errors + "\n")
+
+        fixed_color = _BOLD_RED if diff.num_fixed_errors else _GREEN
```
I did have that in my first iteration, but I came around to doing it this way because I thought that this would be more appropriate when seen in a CI (or pre-commit) context. This matches the fact that the return-code will be non-zero when changes to the ratchet-file are made.

(Although I do see the argument that fixing errors is a good thing, and would usually be green.)

Tenzer replied:
Perhaps making the text yellow when there are fixed errors would be better?

meshy replied:
I can do that!

meshy commented:
@Tenzer I've addressed your points an rebased them in.

The commit which changes is "[Add ColorChangeReportWriter](https://github.com/Memrise/mypy-json-report/pull/89/commits/b76aab2f07cc587630c3bfb19c728b909ca10aff)", and I have created one new commit, "[Change spacing around report summary](https://github.com/Memrise/mypy-json-report/pull/89/commits/c209c3e694f056c3c087a9ed295dbc9f6638bad9)".

Tenzer commented on mypy_json_report.py:
```diff
@@ -229,6 +242,102 @@ class DiffReport:
+                + self._highlight_quotes(error_msg)
+                + code
+            )
+        elif ": note: " in line:
+            location, _, message = line.partition(" note: ")
+            return (
+                location
+                + self._style(self._BLUE, " note: ")
+                + self._highlight_quotes(message)
+            )
+        else:
+            return line
```
**Nit:** Remove unnecessary `else` branches since you are returning already:
```suggestion

        if ": note: " in line:
            location, _, message = line.partition(" note: ")
            return (
                location
                + self._style(self._BLUE, " note: ")
                + self._highlight_quotes(message)
            )

        return line
```

Tenzer replied:
I don't think this was corrected?

Tenzer replied:
Ah, you've done it now. Thanks!

Tenzer reviewed (approved):
One minor thing left, then I think this is good to merge (assuming that you're ready for it - the PR is currently a draft).

I'll make a release when this has been merged in.

meshy commented:
All updated -- thank you!
